File: deep_research/static/js/activity.js
        let activityEvents = [];

        const activityStatusLabels = {
            running: "进行中",
            completed: "已完成",
            failed: "失败",
            cancelled: "已取消",
        };

        function isValidActivityEvent(event) {
            return Boolean(
                event &&
                typeof event === "object" &&
                typeof event.id === "string" &&
                typeof event.timestamp === "string" &&
                typeof event.kind === "string" &&
                typeof event.actor === "string" &&
                typeof event.status === "string" &&
                typeof event.label === "string" &&
                typeof event.summary === "string" &&
                typeof event.detail === "string",
            );
        }

        function activityTimestamp() {
            return new Date().toISOString();
        }

        function activityId() {
            if (
                globalThis.crypto &&
                typeof globalThis.crypto.randomUUID === "function"
            ) {
                return `client-${globalThis.crypto.randomUUID()}`;
            }

            return `client-${Date.now()}-${Math.random().toString(16).slice(2)}`;
        }

        function createClientActivity(
            kind,
            actor,
            status,
            label,
            detail = "",
            elapsedMs = null,
            summary = label,
        ) {
            return {
                id: activityId(),
                run_id: "client",
                timestamp: activityTimestamp(),
                kind,
                actor,
                status,
                label,
                summary,
                detail,
                elapsed_ms: elapsedMs,
            };
        }

        function normalizeActivityEvents(events) {
            if (!Array.isArray(events)) {
                return [];
            }

            const seen = new Set();
            return events
                .filter((event) => {
                    if (!isValidActivityEvent(event) || seen.has(event.id)) {
                        return false;
                    }

                    seen.add(event.id);
                    return true;
                })
                .sort((left, right) => (
                    Date.parse(left.timestamp) - Date.parse(right.timestamp)
                ));
        }

        function formatActivityTime(value) {
            const timestamp = Date.parse(value);
            if (!Number.isFinite(timestamp)) {
                return "时间未知";
            }

            return new Intl.DateTimeFormat("zh-CN", {
                month: "2-digit",
                day: "2-digit",
                hour: "2-digit",
                minute: "2-digit",
                second: "2-digit",
            }).format(timestamp);
        }

        function formatActivityElapsed(value) {
            if (
                typeof value !== "number" ||
                !Number.isFinite(value) ||
                value < 0
            ) {
                return "";
            }

            if (value < 1000) {
                return `${Math.round(value)} ms`;
            }

            return `${(value / 1000).toFixed(1)} s`;
        }

        function renderActivityLog(events = activityEvents) {
            activityEvents = normalizeActivityEvents(events);

            if (!activityList) {
                return;
            }

            activityList.replaceChildren();

            for (const event of activityEvents) {
                const item = document.createElement("li");
                item.className = `activity-item activity-${event.status}`;

                const marker = document.createElement("span");
                marker.className = "activity-marker";
                marker.setAttribute("aria-hidden", "true");

                const body = document.createElement("div");
                body.className = "activity-item-body";

                const heading = document.createElement("div");
                heading.className = "activity-item-heading";

                const label = document.createElement("strong");
                label.className = "activity-item-label";
                label.textContent = event.label;

                const state = document.createElement("span");
                state.className = "activity-item-state";
                state.textContent = (
                    activityStatusLabels[event.status] || event.status
                );
                heading.append(label, state);

                body.append(heading);

                if (event.detail) {
                    const detail = document.createElement("p");
                    detail.className = "activity-item-detail";
                    detail.textContent = event.detail;
                    body.append(detail);
                }

                const meta = document.createElement("div");
                meta.className = "activity-item-meta";
                const elapsed = formatActivityElapsed(event.elapsed_ms);
                meta.textContent = elapsed
                    ? `${formatActivityTime(event.timestamp)} · ${elapsed}`
                    : formatActivityTime(event.timestamp);
                body.append(meta);

                item.append(marker, body);
                activityList.append(item);
            }

            if (activityStatus) {
                activityStatus.textContent = activityEvents.length
                    ? `共 ${activityEvents.length} 条记录`
                    : "当前会话还没有研究日志";
            }

            if (activitySummary) {
                activitySummary.textContent = activityEvents.length
                    ? "问题、计划和执行步骤会按时间记录在这里"
                    : "记录当前会话的问题和执行过程";
            }
        }

        function resetActivityLog() {
            activityEvents = [];
            renderActivityLog();
        }

        function setActivityEvents(events) {
            renderActivityLog(events);
        }

        function appendActivityEvent(event) {
            if (!isValidActivityEvent(event)) {
                return;
            }

            renderActivityLog([...activityEvents, event]);
        }

        function activitySummaryForTask(task) {
            const agentName = String(task?.agent_name || "researcher");
            const activity = [...activityEvents]
                .reverse()
                .find(
                    (event) => (
                        event.kind === "agent" &&
                        event.actor === agentName &&
                        event.status === "running" &&
                        event.summary
                    ),
                );

            if (activity) {
                return activity.summary;
            }

            const description = String(task?.description || "");
            const compacted = description.replace(/\s+/g, " ").trim();

            if (compacted.length <= 140) {
                return compacted;
            }

            return `${compacted.slice(0, 139)}…`;
        }

        function recordQuestionActivity(question) {
            appendActivityEvent(
                createClientActivity(
                    "question",
                    "user",
                    "completed",
                    "用户提出问题",
                    question,
                ),
            );
            appendActivityEvent(
                createClientActivity(
                    "research",
                    "main",
                    "running",
                    "开始处理问题",
                ),
            );
        }

        function recordActivityFromStreamEvent(event) {
            if (!event || typeof event.type !== "string") {
                return;
            }

            const type = event.type;
            const agentName = String(event.agent_name || "researcher");
            const name = String(event.name || "unknown_tool");
            let activity = null;

            if (type === "plan_update") {
                const count = Array.isArray(event.todos)
                    ? event.todos.length
                    : 0;
                activity = createClientActivity(
                    "plan",
                    "main",
                    "completed",
                    "研究计划已更新",
                    `当前包含 ${count} 个步骤`,
                );
            } else if (type === "subagent_start") {
                activity = createClientActivity(
                    "agent",
                    agentName,
                    "running",
                    `${agentName} 开始执行任务`,
                    String(event.description || ""),
                );
            } else if (type === "subagent_end") {
                const state = String(event.status || "completed");
                const sourceCount = Array.isArray(event.source_ids)
                    ? event.source_ids.length
                    : 0;
                activity = createClientActivity(
                    "agent",
                    agentName,
                    state,
                    state === "completed"
                        ? `${agentName} 已完成任务`
                        : `${agentName} 任务${activityStatusLabels[state] || state}`,
                    state === "completed"
                        ? `关联 ${sourceCount} 个来源`
                        : String(event.error || "任务未完成"),
                );
            } else if (type === "tool_start") {
                if (name === "prepare_article_for_publication") {
                    activity = createClientActivity(
                        "publishing",
                        "main",
                        "running",
                        "开始整理文章草稿",
                        "正在根据研究 Artifact 准备 Article 草稿",
                    );
                } else {
                    activity = createClientActivity(
                        "tool",
                        "agent",
                        "running",
                        `开始使用 ${name}`,
                    );
                }
            } else if (type === "tool_end") {
                const state = String(event.status || "completed");
                if (name === "prepare_article_for_publication") {
                    const failed = state === "failed";
                    activity = createClientActivity(
                        "publishing",
                        "main",
                        state,
                        failed
                            ? "文章草稿整理失败"
                            : "文章草稿整理完成",
                        failed
                            ? "文章草稿未能准备完成"
                            : "草稿已准备完成，等待用户人工审批",
                        typeof event.elapsed_ms === "number"
                            ? event.elapsed_ms
                            : null,
                    );
                } else {
                    activity = createClientActivity(
                        "tool",
                        "agent",
                        state,
                        `${name} 执行结束`,
                        "",
                        typeof event.elapsed_ms === "number"
                            ? event.elapsed_ms
                            : null,
                    );
                }
            } else if (type === "memory_compacted") {
                activity = createClientActivity(
                    "memory",
                    "system",
                    "completed",
                    "上下文已整理",
                );
            } else if (type === "memory_recall_started" || type === "memory_recall_completed") {
                activity = createClientActivity(
                    "memory",
                    "system",
                    type === "memory_recall_started" ? "running" : "completed",
                    type === "memory_recall_started" ? "正在检索长期记忆" : "长期记忆检索完成",
                );
            } else if (type === "model_started" || type === "model_reasoning" || type === "model_retrying" || type === "model_completed" || type === "model_failed" || type === "heartbeat") {
                const failed = type === "model_failed";
                const completed = type === "model_completed";
                activity = createClientActivity(
                    "model",
                    String(event.agent_name || "main"),
                    failed ? "failed" : completed ? "completed" : "running",
                    type === "model_retrying"
                        ? `模型响应异常，正在进行第 ${event.attempt || "?"} 次重试`
                        : type === "model_reasoning"
                            ? String(event.message || "正在分析请求")
                            : type === "heartbeat"
                                ? "仍在处理中"
                                : failed
                                    ? "模型调用失败"
                                    : completed
                                        ? "模型阶段完成"
                                        : "正在分析请求",
                );
            } else if (type === "artifact_saved") {
                activity = createClientActivity(
                    "artifact",
                    "agent",
                    "completed",
                    "研究报告已保存",
                    String(event.filename || ""),
                );
            } else if (type === "done") {
                const goalStatus = event.goal_status || "completed";
                activity = createClientActivity(
                    "research",
                    "main",
                    goalStatus,
                    goalStatus === "waiting_for_user"
                        ? "等待用户确认"
                        : goalStatus === "completed_with_failure"
                            ? "处理结束，但目标未完成"
                            : "处理完成",
                    Array.isArray(event.error_codes)
                        ? event.error_codes.join(", ")
                        : "",
                );
            } else if (type === "cancelled") {
                activity = createClientActivity(
                    "research",
                    "main",
                    "cancelled",
                    "研究已取消",
                );
            } else if (type === "cancel_requested") {
                activity = createClientActivity(
                    "research",
                    "main",
                    "running",
                    "正在取消研究",
                );
            } else if (type === "error") {
                activity = createClientActivity(
                    "research",
                    "main",
                    "failed",
                    "研究失败",
                    String(event.error || "研究服务发生错误"),
                );
            }

            if (activity) {
                appendActivityEvent(activity);
            }
        }

        function openActivityDialog() {
            renderActivityLog();

            if (typeof activityDialog?.showModal === "function") {
                activityDialog.showModal();
            } else {
                activityDialog?.setAttribute("open", "");
            }
        }

        function closeActivityDialog() {
            if (typeof activityDialog?.close === "function") {
                activityDialog.close();
            } else {
                activityDialog?.removeAttribute("open");
            }
        }

        activityOpenButton?.addEventListener("click", openActivityDialog);
        activityCloseButton?.addEventListener("click", closeActivityDialog);

        renderActivityLog();

File: deep_research/activity.py
from datetime import datetime, timezone
from uuid import uuid4

from typing_extensions import TypedDict


MAX_ACTIVITY_EVENTS = 200


class ActivityEvent(TypedDict):
    id: str # 核心标识字段
    run_id: str # 关联的运行 ID
    timestamp: str # 时间字段
    kind: str # 分类与来源字段
    actor: str 
    status: str # 状态与展示字段
    label: str
    summary: str
    detail: str
    elapsed_ms: float | None # 性能字段

# 时间
def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace(
        "+00:00",
        "Z",
    )

# 文本压缩与截断工具
def compact_text(value: str, *, max_length: int = 140) -> str:
    compacted = " ".join(value.split())

    if len(compacted) <= max_length:
        return compacted

    return f"{compacted[: max_length - 1]}…"


def build_activity_event(
    *,
    run_id: str, # 关联的运行 ID
    kind: str, # 分类与来源字段
    actor: str,
    status: str, # 状态与展示字段
    label: str,
    summary: str | None = None,
    detail: str = "",
    elapsed_ms: float | None = None, # 性能字段
) -> ActivityEvent:
    return {
        "id": uuid4().hex,
        "run_id": run_id,
        "timestamp": utc_now(),
        "kind": kind,
        "actor": actor,
        "status": status,
        "label": label,
        "summary": summary if summary is not None else label,
        "detail": detail,
        "elapsed_ms": elapsed_ms,
    }


def build_question_activity(
    question: str,
    *,
    run_id: str,
) -> ActivityEvent:
    return build_activity_event(
        run_id=run_id,
        kind="question",
        actor="user",
        status="completed",
        label="用户提出问题",
        summary=compact_text(question.strip()),
        detail=question.strip(),
    )

# 流式响应中的各种内部事件翻译成结构化的 ActivityEvent 对象，为前端提供统一的活动时间线/进度展示数据
def activity_from_stream_event(
    event: dict[str, object],
    *,
    run_id: str,
) -> ActivityEvent | None:
    event_type = event.get("type")

    if event_type == "plan_update":
        todos = event.get("todos")
        count = len(todos) if isinstance(todos, list) else 0
        return build_activity_event(
            run_id=run_id,
            kind="plan",
            actor="main",
            status="completed",
            label="研究计划已更新",
            detail=f"当前包含 {count} 个步骤",
        )

    if event_type == "subagent_start":
        agent_name = str(event.get("agent_name") or "researcher")
        description = str(event.get("description") or "")
        return build_activity_event(
            run_id=run_id,
            kind="agent",
            actor=agent_name,
            status="running",
            label=f"已分配给 {agent_name} 任务",
            summary=compact_text(description) or "等待任务描述",
            detail=description,
        )

    if event_type == "subagent_end":
        agent_name = str(event.get("agent_name") or "researcher")
        status = str(event.get("status") or "completed")
        source_ids = event.get("source_ids")
        source_count = (
            len(source_ids)
            if isinstance(source_ids, list)
            else 0
        )
        detail = (
            f"关联 {source_count} 个来源"
            if status == "completed"
            else str(event.get("error") or "任务未完成")
        )
        return build_activity_event(
            run_id=run_id,
            kind="agent",
            actor=agent_name,
            status=status,
            label=(
                f"{agent_name} 已完成任务"
                if status == "completed"
                else f"{agent_name} 任务{status}"
            ),
            summary=compact_text(detail),
            detail=detail,
        )

    if event_type in {"tool_start", "tool_end"}:
        name = str(event.get("name") or "unknown_tool")
        status = (
            "running"
            if event_type == "tool_start"
            else str(event.get("status") or "completed")
        )
        elapsed_ms = event.get("elapsed_ms")

        if name == "prepare_article_for_publication":
            if event_type == "tool_start":
                label = "开始整理文章草稿"
                detail = "正在根据研究 Artifact 准备 Article 草稿"
            elif status == "failed":
                label = "文章草稿整理失败"
                detail = "文章草稿未能准备完成"
            else:
                label = "文章草稿整理完成"
                detail = "草稿已准备完成，等待用户人工审批"

            return build_activity_event(
                run_id=run_id,
                kind="publishing",
                actor="main",
                status=status,
                label=label,
                detail=detail,
                elapsed_ms=(
                    float(elapsed_ms)
                    if isinstance(elapsed_ms, (int, float))
                    and not isinstance(elapsed_ms, bool)
                    else None
                ),
            )

        return build_activity_event(
            run_id=run_id,
            kind="tool",
            actor="agent",
            status=status,
            label=(
                f"开始使用 {name}"
                if event_type == "tool_start"
                else f"{name} 执行结束"
            ),
            elapsed_ms=(
                float(elapsed_ms)
                if isinstance(elapsed_ms, (int, float))
                and not isinstance(elapsed_ms, bool)
                else None
            ),
        )

    if event_type == "memory_compacted":
        return build_activity_event(
            run_id=run_id,
            kind="memory",
            actor="system",
            status="completed",
            label="上下文已整理",
        )

    if event_type == "artifact_saved":
        filename = str(event.get("filename") or "")
        return build_activity_event(
            run_id=run_id,
            kind="artifact",
            actor="agent",
            status="completed",
            label="研究报告已保存",
            detail=filename,
        )

    if event_type == "done":
        goal_status = str(event.get("goal_status") or "completed")
        if goal_status == "waiting_for_user":
            return build_activity_event(
                run_id=run_id,
                kind="research",
                actor="main",
                status="waiting_for_user",
                label="等待用户确认",
            )
        if goal_status == "completed_with_failure":
            error_codes = event.get("error_codes")
            detail = (
                ", ".join(str(code) for code in error_codes)
                if isinstance(error_codes, list)
                else ""
            )
            return build_activity_event(
                run_id=run_id,
                kind="research",
                actor="main",
                status="completed_with_failure",
                label="处理结束，但目标未完成",
                detail=detail,
            )
        return build_activity_event(
            run_id=run_id,
            kind="research",
            actor="main",
            status="completed",
            label="处理完成",
        )

    if event_type == "cancelled":
        return build_activity_event(
            run_id=run_id,
            kind="research",
            actor="main",
            status="cancelled",
            label="研究已取消",
        )

    if event_type == "error":
        return build_activity_event(
            run_id=run_id,
            kind="research",
            actor="main",
            status="failed",
            label="研究失败",
        )

    return None


def merge_activity_events(
    current: list[ActivityEvent] | None,
    update: list[ActivityEvent] | None,
) -> list[ActivityEvent]:
    merged: dict[str, ActivityEvent] = {}
    # 把 current 和 update 两个数组合并成一个可迭代序列
    for event in [*(current or []), *(update or [])]:
        if not isinstance(event, dict):
            continue

        event_id = event.get("id")
        timestamp = event.get("timestamp")

        if not isinstance(event_id, str) or not event_id:
            continue

        if not isinstance(timestamp, str) or not timestamp:
            continue
        # 存入字典（去重）
        merged[event_id] = dict(event)
    # 排序
    ordered = sorted(
        merged.values(),
        key=lambda item: (item["timestamp"], item["id"]),
    )
    return ordered[-MAX_ACTIVITY_EVENTS:]
